src/coreclr/src/gc/unix/gcenv.unix.cpp

```diff
@@ -502,6 +522,27 @@ void GCToOSInterface::FlushProcessWriteBuffers()
     }
     else
     {
+#if defined(TARGET_OSX) && defined(TARGET_ARM64)
+        mach_msg_type_number_t cThreads;
+        thread_act_t *pThreads;
+        kern_return_t machret = task_threads(mach_task_self(), &pThreads, &cThreads);
+        CHECK_MACH("task_threads()", machret);
+
+        uintptr_t sp;
+        uintptr_t registerValues[128];
+
+        // Iterate through each of the threads in the list.
+        for (mach_msg_type_number_t i = 0; i < cThreads; i++)
+        {
+            // Request the threads pointer values to force the thread to emit a memory barrier
+            size_t registers = 128;
+            machret = thread_get_register_pointer_values(pThreads[i], &sp, &registers, registerValues);
```

Comment:
When I was experimenting with this stuff, I've also tried to add a check for the thread being swapped or not. For large number of threads, it dramatically reduced the number of threads we need to call the `thread_get_register_pointer_values` for. Here is what I had there:
```
        thread_basic_info info;
        mach_msg_type_number_t infoCount = THREAD_BASIC_INFO_COUNT;
        machret = thread_info(thread, THREAD_BASIC_INFO, (thread_info_t)&info, &infoCount);
        assert(machret == KERN_SUCCESS);

        if (!(info.flags & TH_FLAGS_SWAPPED))
        {
            machret = thread_get_register_pointer_values(pThreads[i], &sp, &registers, registerValues);
        }
```
The idea was that when we find the thread is swapped, its write buffers must have already been flushed, as its context is saved and the thread is not executing. And even if it started to execute right after this check, we wouldn't care about its write buffers, as that can be seen as happening "in future" after the call to the FlushProcessWriteBuffers.

Reply:
So which is faster?  And which is safer?

If a thread is swapped the kernel has its registers in memory, so it seems the `thread_get_register_pointer_values` might be quick.  I would guess its speed would be similar to `thread_info`.  The speed might mostly correlate with actual `registers` vs `THREAD_BASIC_INFO`.  The memcpy taking the bulk of the call time.  For active threads there would be two calls required.  The syscall roundtrip is expensive.

I guess I would lean towards simplicity.  .NET philosophy would probably prefer to go with performance. It might depend on how many threads are in the process as well as how many threads are running. Can/have you collect/ed reasonable data to see which is faster (on Amd64 and AppleSilicon)?

Reply:
> when we find the thread is swapped, its write buffers must have already been flushed

It is probably true, but I can imagine it not always being true.  I guess 
- if we have observed the Swapped flag, the swapped flag itself must have been observed
- for Amd64 that seems like that would be sufficient with the strong ordering of writes
- for Arm64 that would only be true if there was a write barrier before writing the swapped flag
  - It is likely such a barrier is included
  - I don't know enough to assert confirm such a barrier is explicitly required

Reply:
> And even if it started to execute right after this check, we wouldn't care about its write buffers, as that can be seen as happening "in future" after the call to the FlushProcessWriteBuffers.

And it would be doing a GC related task... As opposed to running managed code.

Reply:
I've ran the tests again today and I've found that the added check actually makes it a bit worse. For example, for 100 threads, just calling the `thread_get_register_pointer_values` takes 957us and adding the call to `thread_info` that resulted in the other function being called 13 times instead of 100 times in my test took 999us. 
So it doesn't make sense to add that filtering.

Reply:
Thanks for running the experiment.